src/hooks/login/useWalletConnectV2Login.ts

```diff
@@ -0,0 +1,249 @@
+  invalidTopic = 'Expired connection',
+  sessionExpired = 'Unable to connect to existing session'
+}
+
+export interface WalletConnectV2LoginHookCustomStateType {
+  uriDeepLink: string;
+  walletConnectUri?: string;
+  wcPairings?: PairingTypes.Struct[];
+}
+
+export type WalletConnectV2LoginHookReturnType = [
+  (loginProvider?: boolean) => void,
+  (pairing: PairingTypes.Struct) => Promise<void>,
+  LoginHookGenericStateType,
+  WalletConnectV2LoginHookCustomStateType
+];
+
+export const useWalletConnectV2Login = ({
+  callbackRoute,
+  logoutRoute,
+  token,
+  onLoginRedirect
+}: InitWalletConnectV2Type): WalletConnectV2LoginHookReturnType => {
+  const dispatch = useDispatch();
+
+  const [error, setError] = useState<string>('');
+  const [wcUri, setWcUri] = useState<string>('');
+  const [wcPairings, setWcPairings] = useState<
+    PairingTypes.Struct[] | undefined
+  >([]);
+
+  const provider = getAccountProvider();
+  const walletConnectV2RelayAddress = useSelector(walletConnectV2RelaySelector);
+  const walletConnectV2ProjectId = useSelector(
+    walletConnectV2ProjectIdSelector
+  );
+  const chainId = useSelector(chainIDSelector);
+  const walletConnectDeepLink = useSelector(walletConnectDeepLinkSelector);
+  const isLoggedIn = useSelector(isLoggedInSelector);
+  const providerRef = useRef<any>(provider);
+
+  const hasWcUri = Boolean(wcUri);
+  const isLoading = !hasWcUri;
+  const uriDeepLink = hasWcUri
+    ? `${walletConnectDeepLink}?wallet-connect=${encodeURIComponent(wcUri)}`
+    : '';
+
+  useUpdateEffect(() => {
+    generateWcUri();
+  }, [token]);
+
+  useUpdateEffect(() => {
+    providerRef.current = provider;
+  }, [provider]);
+
+  const handleOnLogout = () => {
+    logout(logoutRoute);
+  };
+
+  async function handleOnLogin() {
+    try {
+      const provider = providerRef.current;
+      if (isLoggedIn) {
+        return;
+      }
+      if (provider == null) {
+        return;
+      }
+      if (Object.keys(provider).length === 0) {
+        return;
+      }
+      const address = await provider.getAddress();
+      const signature = await provider.getSignature();
+      const hasSignature = Boolean(signature);
+      const loginActionData = {
+        address: address,
+        loginMethod: LoginMethodsEnum.walletconnectv2
+      };
+
+      const loginData = {
+        logoutRoute: logoutRoute,
+        loginType: 'walletconnectv2',
+        callbackRoute: callbackRoute ?? window.location.href
+      };
+
+      if (hasSignature) {
+        dispatch(setWalletConnectLogin(loginData));
+        dispatch(setTokenLoginSignature(signature));
+      } else {
+        dispatch(setWalletConnectLogin(loginData));
+      }
+
+      dispatch(loginAction(loginActionData));
+      optionalRedirect(callbackRoute, onLoginRedirect);
+    } catch (err) {
+      setError(WalletConnectV2Error.invalidAddress);
+      console.error(err);
+    }
+  }
+
+  async function initiateLogin(loginProvider = true) {
```

Comment:
there are many functions in this file. You could create a folder with index, types and other functions to break it into smaller pieces

Reply:
kept them all in the same file as the other providers to keep consistency